File: openlife/RustServer/crates/ol-web/src/ops_dashboard.rs
//! Ops dashboard HTML: duration units, human labels, charts.

use ol_metrics::{CounterSnapshot, OpsSample};

/// Format a duration stored as microseconds.
///
/// Units finer than milliseconds (`µs`) are used only when the value is below 1 ms.
pub fn format_ops_duration_us(us: u64) -> String {
    if us < 1_000 {
        format!("{us} µs")
    } else if us < 1_000_000 {
        if us % 1_000 == 0 {
            format!("{} ms", us / 1_000)
        } else if us % 100 == 0 {
            format!("{:.1} ms", us as f64 / 1_000.0)
        } else {
            format!("{:.2} ms", us as f64 / 1_000.0)
        }
    } else if us % 1_000_000 == 0 {
        format!("{} s", us / 1_000_000)
    } else {
        format!("{:.2} s", us as f64 / 1_000_000.0)
    }
}

/// Format a duration stored as whole milliseconds.
pub fn format_ops_duration_ms(ms: u64) -> String {
    format_ops_duration_us(ms.saturating_mul(1_000))
}

/// UTC `YYYY-MM-DD HH:MM:SS UTC` from Unix milliseconds. `0` → em dash.
pub fn format_unix_ms_utc(ms: u64) -> String {
    if ms == 0 {
        return "—".into();
    }
    let unix = (ms / 1_000) as i64;
    let days = unix.div_euclid(86_400);
    let sod = unix.rem_euclid(86_400) as u64;
    let h = sod / 3_600;
    let m = (sod % 3_600) / 60;
    let s = sod % 60;
    let (y, mo, d) = civil_from_unix_days(days);
    format!("{y:04}-{mo:02}-{d:02} {h:02}:{m:02}:{s:02} UTC")
}

/// Howard Hinnant civil-from-days, `days` since 1970-01-01.
fn civil_from_unix_days(days: i64) -> (i32, u32, u32) {
    let z = days + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = (z - era * 146_097) as u64;
    let yoe = (doe - doe / 1460 + doe / 36_524 - doe / 146_096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if m <= 2 { y + 1 } else { y };
    (y as i32, m as u32, d as u32)
}

fn card(label: &str, tip: &str, field: &str, value: &str, extra_html: &str) -> String {
    let extra = extra_html;
    format!(
        "<div class=\"card\"><strong>{label}<span class=\"tip\">{tip}</span></strong>\
         <br/><span data-field=\"{field}\">{value}</span>{extra}</div>"
    )
}

/// Full `/ops` document. Script is appended without `format!` so JSON `{` is safe.
pub fn build_ops_dashboard_html(
    s: &CounterSnapshot,
    samples: &[OpsSample],
    version: &str,
) -> String {
    let last = samples.last();
    let tick_us = last
        .map(|x| u64::from(x.tick_work_us))
        .unwrap_or(s.tick_work_ema_us);
    let intent_us = last
        .map(|x| u64::from(x.intent_ema_us))
        .unwrap_or(s.intent_ema_us);
    let lock_us = last
        .map(|x| u64::from(x.lock_wait_ema_us))
        .unwrap_or(s.lock_wait_ema_us);

    let mut cards = String::new();
    cards.push_str(&card(
        "Started",
        "When this server process started (UTC).",
        "started",
        &format_unix_ms_utc(s.start_unix_ms),
        "",
    ));
    cards.push_str(&card(
        "Tick",
        "How many simulation ticks have run since start.",
        "tick",
        &s.ticks.to_string(),
        "",
    ));
    cards.push_str(&card(
        "Skip Ticks",
        "Extra simulation steps when the server was behind real time. Catch-up advances, not dropped frames.",
        "skipTicks",
        &s.skip_ticks.to_string(),
        "",
    ));
    cards.push_str(&card(
        "Tick Work",
        "Time spent running one simulation tick (smoothed). Lower is healthier.",
        "tickWork",
        &format_ops_duration_us(tick_us),
        "",
    ));
    cards.push_str(&card(
        "Intent Time",
        "Time to apply one player or NPC action (smoothed average).",
        "intent",
        &format_ops_duration_us(intent_us),
        "",
    ));
    cards.push_str(&card(
        "Tick Average / P90",
        "Average tick work, and the slowest 10% (90th percentile).",
        "tickAvgP90",
        &format!(
            "{} / {}",
            format_ops_duration_us(s.tick_work_avg_us),
            format_ops_duration_us(s.tick_work_p90_us)
        ),
        "",
    ));
    cards.push_str(&card(
        "Tick Outliers / Normal",
        "How many recent ticks were unusually slow versus typical.",
        "tickOutliers",
        &format!("{} / {}", s.tick_work_outliers, s.tick_work_normal),
        "",
    ));
    cards.push_str(&card(
        "Intent Average / P90",
        "Average action-apply time, and the slowest 10%.",
        "intentAvgP90",
        &format!(
            "{} / {}",
            format_ops_duration_us(s.intent_avg_us),
            format_ops_duration_us(s.intent_p90_us)
        ),
        "",
    ));
    cards.push_str(&card(
        "Intent Outliers / Normal",
        "How many recent action applies were unusually slow versus typical.",
        "intentOutliers",
        &format!("{} / {}", s.intent_outliers, s.intent_normal),
        "",
    ));
    cards.push_str(&card(
        "Human Reply Average / P90",
        "Time to apply actions from human clients, average and slowest 10%.",
        "humanIntent",
        &format!(
            "{} / {}",
            format_ops_duration_us(s.human_intent_avg_us),
            format_ops_duration_us(s.human_intent_p90_us)
        ),
        &format!(
            "<br/><span class=\"muted\" data-field=\"humanIntentN\">n = {}</span>",
            s.human_intent_count
        ),
    ));
    cards.push_str(&card(
        "AI Intent Average / P90",
        "Time to apply NPC / AI actions, average and slowest 10%.",
        "aiIntent",
        &format!(
            "{} / {}",
            format_ops_duration_us(s.ai_intent_avg_us),
            format_ops_duration_us(s.ai_intent_p90_us)
        ),
        &format!(
            "<br/><span class=\"muted\" data-field=\"aiIntentN\">n = {}</span>",
            s.ai_intent_count
        ),
    ));
    cards.push_str(&card(
        "Lock Wait",
        "Time spent waiting for the world lock before a tick can run.",
        "lockWait",
        &format_ops_duration_us(lock_us),
        "",
    ));
    cards.push_str(&card(
        "Boot Total",
        "How long the server took to start (objects, transitions, and world load).",
        "bootTotal",
        &format_ops_duration_ms(s.boot_total_ms),
        "",
    ));
    cards.push_str(&card(
        "Boot Objects / Transitions / World",
        "Start-up time split: object data, transitions, then the world map.",
        "bootParts",
        &format!(
            "{} / {} / {}",
            format_ops_duration_ms(s.boot_objects_ms),
            format_ops_duration_ms(s.boot_transitions_ms),
            format_ops_duration_ms(s.boot_world_ms)
        ),
        "",
    ));
    cards.push_str(&card(
        "Samples",
        "How many timing samples are currently kept for the graphs.",
        "samples",
        &samples.len().to_string(),
        "",
    ));
    cards.push_str(&card(
        "AI Thinks",
        "How many NPC think steps have run since start.",
        "aiThinks",
        &s.ai_thinks.to_string(),
        "",
    ));

    let metrics = crate::metrics_json_from_snapshot(s, version, serde_json::json!({}));
    let series = crate::ops_series_json(samples);
    let metrics_js = serde_json::to_string(&metrics).unwrap_or_else(|_| "{}".into());
    let series_js = serde_json::to_string(&series).unwrap_or_else(|_| "{\"samples\":[],\"count\":0}".into());

    let mut html = String::with_capacity(24_000 + cards.len() + metrics_js.len() + series_js.len());
    html.push_str(OPS_HEAD);
    html.push_str(&cards);
    html.push_str(OPS_CHARTS);
    html.push_str("<script>window.OLR_OPS_METRICS=");
    html.push_str(&metrics_js);
    html.push_str(";window.OLR_OPS=");
    html.push_str(&series_js);
    html.push_str(";</script>\n<script>");
    html.push_str(include_str!("ops_dashboard.js"));
    html.push_str("</script>\n</body></html>");
    html
}

const OPS_HEAD: &str = r##"<!DOCTYPE html>
<html lang="en"><head><meta charset="utf-8"/><title>Ops — Open Life Reborn</title>
<style>
body{font-family:system-ui,sans-serif;background:#0b0f14;color:#e7eef7;margin:2rem;line-height:1.5}
a{color:#6ec6ff}
.cards{display:flex;flex-wrap:wrap;gap:1rem}
.card{background:#1a222e;padding:1rem 1.25rem;border-radius:8px;min-width:11rem;position:relative}
.card strong{position:relative;cursor:help;border-bottom:1px dotted #6ec6ff}
.card .tip,.chart-head h3 .tip{
  display:none;position:absolute;left:0;top:calc(100% + .35rem);z-index:5;
  background:#0e1620;color:#e7eef7;border:1px solid #2a3a4e;border-radius:6px;
  padding:.55rem .7rem;width:16rem;font-weight:400;font-size:.85rem;line-height:1.35;
  box-shadow:0 8px 24px rgba(0,0,0,.4)
}
.card:hover .tip,.chart-head h3:hover .tip{display:block}
.muted{color:#8b9bb0;font-size:.9rem}
.chart{margin:1.5rem 0;background:#121a24;border:1px solid #1e2a3a;border-radius:8px;padding:1rem}
.chart-head{display:flex;flex-wrap:wrap;align-items:center;justify-content:space-between;gap:.75rem;margin-bottom:.5rem}
.chart-head h3{margin:0;font-size:1.05rem;position:relative;cursor:help}
.chart-head label,.time-display{color:#8b9bb0;font-size:.9rem}
.time-display select,.chart-head select{background:#0b0f14;color:#e7eef7;border:1px solid #2a3a4e;border-radius:4px;padding:.25rem .4rem}
.chart-wrap{position:relative}
.chart-wrap svg{width:100%;height:180px;display:block;background:#0b0f14;border-radius:4px}
.chart-wrap .line{stroke:#6ec6ff;stroke-width:2;fill:none}
.chart-wrap .grid{stroke:#1e2a3a;stroke-width:1}
.chart-wrap .axis{stroke:#3a4a5e;stroke-width:1}
.chart-wrap .axis-label{fill:#8b9bb0;font-size:11px;font-family:system-ui,sans-serif}
.chart-wrap .cross{stroke:#e7eef7;stroke-width:1;stroke-dasharray:3 3;opacity:.7}
.chart-wrap .dot{fill:#6ec6ff;stroke:#0b0f14;stroke-width:1}
.hover-tip{
  position:absolute;pointer-events:none;background:#0e1620;color:#e7eef7;
  border:1px solid #2a3a4e;border-radius:6px;padding:.45rem .6rem;font-size:.85rem;
  line-height:1.35;min-width:10rem;box-shadow:0 8px 24px rgba(0,0,0,.45)
}
</style></head><body>
<p><a href="/">← Home</a></p>
<h1>Ops Dashboard</h1>
<p class="muted">Skip Ticks are catch-up advances when the server lags, not dropped wakes. Samples stay in memory about 5 seconds and flush about every 5 minutes. Graph time range is saved in this browser.</p>
<div class="cards">
"##;

const OPS_CHARTS: &str = r##"</div>
<p class="muted">Latency: average plus the slowest about 10% (P90), and outlier versus normal counts. Boot timings are recorded once at server start. Hover a graph for the value and time under the pointer.</p>
<p class="time-display"><label>Time range
<select id="ops-time-range" title="How much history the graphs show">
<option value="all">All samples</option>
<option value="30s">Last 30 seconds</option>
<option value="1m">Last 1 minute</option>
<option value="5m" selected>Last 5 minutes</option>
<option value="15m">Last 15 minutes</option>
<option value="1h">Last 1 hour</option>
</select></label></p>
<div id="ops-charts"></div>
<p class="muted"><a href="/api/ops/series">JSON series</a> · <a href="/api/metrics">JSON metrics</a></p>
"##;

#[cfg(test)]
mod tests {
    use super::*;
    use ol_metrics::Counters;

    #[test]
    fn duration_us_only_below_one_ms() {
        assert_eq!(format_ops_duration_us(0), "0 µs");
        assert_eq!(format_ops_duration_us(1), "1 µs");
        assert_eq!(format_ops_duration_us(999), "999 µs");
        assert_eq!(format_ops_duration_us(1_000), "1 ms");
        assert_eq!(format_ops_duration_us(1_500), "1.5 ms");
        assert_eq!(format_ops_duration_us(10_000), "10 ms");
        assert_eq!(format_ops_duration_us(1_000_000), "1 s");
        assert_eq!(format_ops_duration_us(1_500_000), "1.50 s");
    }

    #[test]
    fn duration_ms_promotes_to_seconds() {
        assert_eq!(format_ops_duration_ms(0), "0 µs");
        assert_eq!(format_ops_duration_ms(1), "1 ms");
        assert_eq!(format_ops_duration_ms(500), "500 ms");
        assert_eq!(format_ops_duration_ms(1_200), "1.20 s");
    }

    #[test]
    fn unix_epoch_and_one_second() {
        assert_eq!(format_unix_ms_utc(0), "—");
        assert_eq!(format_unix_ms_utc(1_000), "1970-01-01 00:00:01 UTC");
    }

    #[test]
    fn dashboard_labels_have_no_underscores() {
        let s = Counters::new().snapshot();
        let html = build_ops_dashboard_html(&s, &[], "0.1.0");
        assert!(html.contains("Skip Ticks"));
        assert!(html.contains("Tick Work"));
        assert!(html.contains("Intent Time"));
        assert!(html.contains("Lock Wait"));
        assert!(html.contains("Boot Total"));
        assert!(html.contains("class=\"tip\""));
        assert!(!html.contains("<strong>skip_ticks"));
        assert!(!html.contains(">tick_work_us<"));
        assert!(!html.contains("http-equiv=\"refresh\""));
        assert!(html.contains("localStorage"));
        assert!(html.contains("hover-tip"));
        assert!(html.contains("ops-time-range"));
        assert!(html.contains("Time range"));
        assert!(!html.contains("data-scale"));
    }
}

File: openlife/RustServer/crates/ol-web/src/ops_dashboard.js
(function () {
  "use strict";

  var STORAGE_KEY = "openlife-ops-dashboard-v2";

  var TIME_RANGES = {
    all: 0,
    "30s": 30 * 1000,
    "1m": 60 * 1000,
    "5m": 5 * 60 * 1000,
    "15m": 15 * 60 * 1000,
    "1h": 60 * 60 * 1000,
    "1d": 24 * 60 * 60 * 1000,
    "1w": 7 * 24 * 60 * 60 * 1000,
    "30d": 30 * 24 * 60 * 60 * 1000
  };

  var CHARTS = [
    {
      id: "tickWork",
      title: "Tick Work",
      tip: "Time spent running one simulation tick, smoothed. Lower is healthier.",
      field: "tick_work_us",
      kind: "duration"
    },
    {
      id: "intent",
      title: "Intent Time",
      tip: "Time to apply one player or NPC action (smoothed average).",
      field: "intent_ema_us",
      kind: "duration"
    },
    {
      id: "lockWait",
      title: "Lock Wait",
      tip: "Time spent waiting for the world lock before a tick can run.",
      field: "lock_wait_ema_us",
      kind: "duration"
    },
    {
      id: "skipTicks",
      title: "Skip Ticks",
      tip: "Catch-up steps when the server fell behind real time. These are extra advances, not dropped frames.",
      field: "skip_ticks",
      kind: "count"
    },
    {
      id: "humanIntent",
      title: "Human Intent Time",
      tip: "Time to apply one human client action (smoothed).",
      field: "human_intent_avg_us",
      kind: "duration"
    },
    {
      id: "simCpu",
      title: "Sim CPU Total",
      tip: "Cumulative simulation tick work.",
      field: "sim_cpu_us",
      kind: "duration"
    }
  ];

  function loadSettings() {
    try {
      var raw = localStorage.getItem(STORAGE_KEY);
      if (!raw) return {};
      var parsed = JSON.parse(raw);
      return parsed && typeof parsed === "object" ? parsed : {};
    } catch (e) {
      return {};
    }
  }

  function saveSettings(s) {
    try {
      localStorage.setItem(STORAGE_KEY, JSON.stringify(s));
    } catch (e) {
      /* private mode */
    }
  }

  function formatDurationUs(us) {
    us = Number(us) || 0;
    if (us < 1000) return Math.round(us) + " µs";
    if (us < 1000000) {
      if (us % 1000 === 0) return us / 1000 + " ms";
      if (us % 100 === 0) return (us / 1000).toFixed(1) + " ms";
      return (us / 1000).toFixed(2) + " ms";
    }
    if (us % 1000000 === 0) return us / 1000000 + " s";
    return (us / 1000000).toFixed(2) + " s";
  }

  function formatDurationMs(ms) {
    return formatDurationUs((Number(ms) || 0) * 1000);
  }

  function formatUnixMs(ms) {
    ms = Number(ms) || 0;
    if (!ms) return "—";
    var d = new Date(ms);
    if (isNaN(d.getTime())) return "—";
    function pad(n) {
      return n < 10 ? "0" + n : String(n);
    }
    return (
      d.getUTCFullYear() +
      "-" +
      pad(d.getUTCMonth() + 1) +
      "-" +
      pad(d.getUTCDate()) +
      " " +
      pad(d.getUTCHours()) +
      ":" +
      pad(d.getUTCMinutes()) +
      ":" +
      pad(d.getUTCSeconds()) +
      " UTC"
    );
  }

  function formatCount(n) {
    n = Number(n) || 0;
    return String(Math.round(n));
  }

  function formatValue(kind, v) {
    if (kind === "duration") return formatDurationUs(v);
    if (kind === "percent") return Math.round(Number(v) || 0) + "%";
    return formatCount(v);
  }

  var settings = loadSettings();

  function currentTimeRange() {
    var id = settings.timeRange || "5m";
    return TIME_RANGES.hasOwnProperty(id) ? id : "5m";
  }

  function filterSamples(samples) {
    samples = samples || [];
    var id = currentTimeRange();
    var windowMs = TIME_RANGES[id] || 0;
    if (!windowMs || samples.length === 0) return samples;
    var latest = Number(samples[samples.length - 1].wall_unix_ms) || Date.now();
    var cut = latest - windowMs;
    var out = [];
    for (var i = 0; i < samples.length; i++) {
      if (Number(samples[i].wall_unix_ms) >= cut) out.push(samples[i]);
    }
    return out;
  }

  function yMaxFor(values) {
    var m = 1;
    for (var i = 0; i < values.length; i++) {
      if (values[i] > m) m = values[i];
    }
    return m;
  }

  function $(sel, root) {
    return (root || document).querySelector(sel);
  }

  function renderCards(metrics, samples) {
    var last = samples.length ? samples[samples.length - 1] : null;
    var lat = (metrics && metrics.latency) || {};
    var boot = (metrics && metrics.boot) || {};
    var set = function (name, text) {
      var el = document.querySelector('[data-field="' + name + '"]');
      if (el) el.textContent = text;
    };
    set("started", formatUnixMs(metrics.start_unix_ms));
    set("tick", formatCount(metrics.ticks));
    set("skipTicks", formatCount(metrics.skip_ticks));
    set(
      "tickWork",
      formatDurationUs(
        last ? last.tick_work_us : metrics.tick_work_ema_us
      )
    );
    set(
      "intent",
      formatDurationUs(last ? last.intent_ema_us : metrics.intent_ema_us)
    );
    set(
      "tickAvgP90",
      formatDurationUs(lat.tick_avg_us) + " / " + formatDurationUs(lat.tick_p90_us)
    );
    set(
      "tickOutliers",
      formatCount(lat.tick_outliers) + " / " + formatCount(lat.tick_normal)
    );
    set(
      "intentAvgP90",
      formatDurationUs(lat.intent_avg_us) +
        " / " +
        formatDurationUs(lat.intent_p90_us)
    );
    set(
      "intentOutliers",
      formatCount(lat.intent_outliers) + " / " + formatCount(lat.intent_normal)
    );
    set(
      "humanIntent",
      formatDurationUs(lat.human_intent_avg_us) +
        " / " +
        formatDurationUs(lat.human_intent_p90_us)
    );
    set("humanIntentN", "n = " + formatCount(lat.human_intent_count));
    set(
      "aiIntent",
      formatDurationUs(lat.ai_intent_avg_us) +
        " / " +
        formatDurationUs(lat.ai_intent_p90_us)
    );
    set("aiIntentN", "n = " + formatCount(lat.ai_intent_count));
    set(
      "lockWait",
      formatDurationUs(last ? last.lock_wait_ema_us : metrics.lock_wait_ema_us)
    );
    set("bootTotal", formatDurationMs(boot.total_ms));
    set(
      "bootParts",
      formatDurationMs(boot.objects_ms) +
        " / " +
        formatDurationMs(boot.transitions_ms) +
        " / " +
        formatDurationMs(boot.world_ms)
    );
    set("samples", formatCount(samples.length));
    set("aiThinks", formatCount(metrics.ai_thinks));
    set("aiThinkTime", formatDurationUs(metrics.ai_think_ema_us));
    set("aiThinkLast", "last " + formatDurationUs(metrics.ai_think_last_us));
    set("aiCpu", formatDurationUs(metrics.ai_cpu_us));
    var tickUs = last ? last.tick_work_us : metrics.tick_work_ema_us;
    var thinkUs = metrics.ai_think_ema_us || 0;
    set(
      "aiVsTick",
      tickUs
        ? Math.min(999, Math.round((thinkUs / tickUs) * 100)) + "%"
        : "—"
    );
    set("aiScanTime", formatDurationUs(metrics.ai_scan_ema_us));
    set("aiScanLast", "last " + formatDurationUs(metrics.ai_scan_last_us));
    set("aiOtherTime", formatDurationUs(metrics.ai_other_ema_us));
    set("aiScanCpu", formatDurationUs(metrics.ai_scan_cpu_us));
    var scanUs = metrics.ai_scan_ema_us || 0;
    set(
      "aiScanVsThink",
      thinkUs
        ? Math.min(100, Math.round((scanUs / thinkUs) * 100)) + "%"
        : "—"
    );
    set(
      "aiScanCache",
      formatCount(metrics.ai_scan_hits) + " / " + formatCount(metrics.ai_scan_calls)
    );
    set("simCpu", formatDurationUs(metrics.sim_cpu_us));
    var simUs = metrics.sim_cpu_us || 0;
    var aiCpu = metrics.ai_cpu_us || 0;
    set(
      "aiVsSim",
      simUs ? Math.min(999, Math.round((aiCpu / simUs) * 100)) + "%" : "—"
    );
  }

  function bindTimeRangeSelect() {
    var sel = document.getElementById("ops-time-range");
    if (!sel || sel.getAttribute("data-bound") === "1") return;
    sel.value = currentTimeRange();
    sel.addEventListener("change", function () {
      settings.timeRange = sel.value;
      saveSettings(settings);
      drawAll(window.OLR_OPS && window.OLR_OPS.samples ? window.OLR_OPS.samples : []);
    });
    sel.setAttribute("data-bound", "1");
  }

  function ensureChartShells() {
    bindTimeRangeSelect();
    var host = document.getElementById("ops-charts");
    if (!host || host.getAttribute("data-ready") === "1") return;
    var html = "";
    for (var i = 0; i < CHARTS.length; i++) {
      var c = CHARTS[i];
      html +=
        '<div class="chart" id="chart-' +
        c.id +
        '">' +
        '<div class="chart-head">' +
        "<h3 title=\"" +
        c.tip.replace(/"/g, "&quot;") +
        '">' +
        c.title +
        '<span class="tip">' +
        c.tip +
        "</span></h3>" +
        "</div>" +
        '<div class="chart-wrap">' +
        '<svg viewBox="0 0 640 180" preserveAspectRatio="none" role="img" aria-label="' +
        c.title +
        '"></svg>' +
        '<div class="hover-tip" hidden></div>' +
        "</div></div>";
    }
    host.innerHTML = html;
    host.setAttribute("data-ready", "1");
  }

  function drawAll(samples) {
    var shown = filterSamples(samples || []);
    for (var i = 0; i < CHARTS.length; i++) drawChart(CHARTS[i], shown);
  }

  function sampleValue(sample, field) {
    var v = sample[field];
    return typeof v === "number" ? v : 0;
  }

  function drawChart(chart, samples) {
    var root = document.getElementById("chart-" + chart.id);
    if (!root) return;
    var svg = root.querySelector("svg");
    var tip = root.querySelector(".hover-tip");
    var W = 640;
    var H = 180;
    var padL = 72;
    var padR = 16;
    var padT = 12;
    var padB = 28;
    var innerW = W - padL - padR;
    var innerH = H - padT - padB;
    var values = [];
    for (var i = 0; i < samples.length; i++) {
      values.push(sampleValue(samples[i], chart.field));
    }
    var yMax = yMaxFor(values);
    if (yMax < 1) yMax = 1;

    var pts = [];
    var n = values.length;
    for (var j = 0; j < n; j++) {
      var x = padL + (n > 1 ? (j / (n - 1)) * innerW : 0);
      var y = padT + innerH - (Math.min(values[j], yMax) / yMax) * innerH;
      pts.push(x.toFixed(1) + "," + y.toFixed(1));
    }

    var y0 = formatValue(chart.kind, 0);
    var yMid = formatValue(chart.kind, yMax / 2);
    var yTop = formatValue(chart.kind, yMax);
    var tFirst = samples.length ? formatUnixMs(samples[0].wall_unix_ms) : "";
    var tLast = samples.length ? formatUnixMs(samples[samples.length - 1].wall_unix_ms) : "";

    var gridY1 = padT + innerH / 2;
    var axisY = padT + innerH;
    var html = "";
    html +=
      '<line class="grid" x1="' +
      padL +
      '" y1="' +
      padT +
      '" x2="' +
      (W - padR) +
      '" y2="' +
      padT +
      '"/>';
    html +=
      '<line class="grid" x1="' +
      padL +
      '" y1="' +
      gridY1 +
      '" x2="' +
      (W - padR) +
      '" y2="' +
      gridY1 +
      '"/>';
    html +=
      '<line class="axis" x1="' +
      padL +
      '" y1="' +
      padT +
      '" x2="' +
      padL +
      '" y2="' +
      axisY +
      '"/>';
    html +=
      '<line class="axis" x1="' +
      padL +
      '" y1="' +
      axisY +
      '" x2="' +
      (W - padR) +
      '" y2="' +
      axisY +
      '"/>';
    html +=
      '<text class="axis-label" x="8" y="' +
      (padT + 4) +
      '">' +
      yTop +
      "</text>";
    html +=
      '<text class="axis-label" x="8" y="' +
      (gridY1 + 4) +
      '">' +
      yMid +
      "</text>";
    html +=
      '<text class="axis-label" x="8" y="' +
      (axisY + 4) +
      '">' +
      y0 +
      "</text>";
    if (tFirst) {
      html +=
        '<text class="axis-label" x="' +
        padL +
        '" y="' +
        (H - 8) +
        '">' +
        tFirst +
        "</text>";
    }
    if (tLast) {
      html +=
        '<text class="axis-label" text-anchor="end" x="' +
        (W - padR) +
        '" y="' +
        (H - 8) +
        '">' +
        tLast +
        "</text>";
    }
    if (pts.length) {
      html +=
        '<polyline class="line" fill="none" points="' + pts.join(" ") + '"/>';
    } else {
      html +=
        '<text class="axis-label" x="' +
        (padL + 12) +
        '" y="' +
        (padT + innerH / 2) +
        '">No samples yet</text>';
    }
    html +=
      '<line class="cross" x1="0" y1="' +
      padT +
      '" x2="0" y2="' +
      axisY +
      '" visibility="hidden"/>';
    html += '<circle class="dot" r="4" visibility="hidden"/>';
    html +=
      '<rect class="hit" x="' +
      padL +
      '" y="' +
      padT +
      '" width="' +
      innerW +
      '" height="' +
      innerH +
      '" fill="transparent"/>';
    svg.innerHTML = html;

    var hit = svg.querySelector(".hit");
    var cross = svg.querySelector(".cross");
    var dot = svg.querySelector(".dot");
    function hideHover() {
      cross.setAttribute("visibility", "hidden");
      dot.setAttribute("visibility", "hidden");
      tip.hidden = true;
    }
    function showAt(clientX) {
      if (!n) {
        hideHover();
        return;
      }
      var rect = svg.getBoundingClientRect();
      var xPix = ((clientX - rect.left) / rect.width) * W;
      var t = (xPix - padL) / innerW;
      if (t < 0) t = 0;
      if (t > 1) t = 1;
      var idx = n > 1 ? Math.round(t * (n - 1)) : 0;
      var sm = samples[idx];
      var val = values[idx];
      var x = padL + (n > 1 ? (idx / (n - 1)) * innerW : 0);
      var y = padT + innerH - (Math.min(val, yMax) / yMax) * innerH;
      cross.setAttribute("x1", x.toFixed(1));
      cross.setAttribute("x2", x.toFixed(1));
      cross.setAttribute("visibility", "visible");
      dot.setAttribute("cx", x.toFixed(1));
      dot.setAttribute("cy", y.toFixed(1));
      dot.setAttribute("visibility", "visible");
      tip.hidden = false;
      tip.innerHTML =
        "<strong>" +
        chart.title +
        "</strong><br/>" +
        formatValue(chart.kind, val) +
        "<br/>" +
        formatUnixMs(sm.wall_unix_ms) +
        "<br/>Tick " +
        formatCount(sm.tick);
      var wrap = root.querySelector(".chart-wrap").getBoundingClientRect();
      var left = clientX - wrap.left + 12;
      var top = y * (rect.height / H) - 8;
      if (left + 200 > wrap.width) left = clientX - wrap.left - 210;
      if (top < 0) top = 0;
      tip.style.left = left + "px";
      tip.style.top = top + "px";
    }
    hit.addEventListener("mousemove", function (ev) {
      showAt(ev.clientX);
    });
    hit.addEventListener("mouseleave", hideHover);
  }

  function apply(metrics, payload) {
    var samples = (payload && payload.samples) || [];
    window.OLR_OPS = payload || { samples: [], count: 0 };
    renderCards(metrics || {}, samples);
    ensureChartShells();
    drawAll(samples);
  }

  function poll() {
    Promise.all([
      fetch("/api/metrics").then(function (r) {
        return r.json();
      }),
      fetch("/api/ops/series").then(function (r) {
        return r.json();
      })
    ])
      .then(function (pair) {
        apply(pair[0], pair[1]);
      })
      .catch(function () {
        /* keep last view */
      });
  }

  apply(
    window.OLR_OPS_METRICS || {},
    window.OLR_OPS || { samples: [], count: 0 }
  );
  setInterval(poll, 5000);
})();
